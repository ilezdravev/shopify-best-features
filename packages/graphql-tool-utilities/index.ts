import {compileToIR} from 'apollo-codegen/lib/compilation');

export const compile = compileToIR;
