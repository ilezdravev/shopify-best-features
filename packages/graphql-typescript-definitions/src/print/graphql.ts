import {
  GraphQLList,
  GraphQLNonNull,
  GraphQLEnumType,
  GraphQLInputObjectType,
  GraphQLInputType,
  GraphQLString,
  GraphQLInt,
  GraphQLFloat,
  GraphQLBoolean,
  GraphQLID,
  GraphQLType,
  GraphQLScalarType,
  GraphQLObjectType,
  GraphQLInterfaceType,
  GraphQLUnionType,
  GraphQLLeafType,
} from 'graphql';

import CodeGenerator from './generator';
import {printPropertyKey, printInterface} from './language';

export type GraphQLPrintableType = GraphQLLeafType | GraphQLObjectType | GraphQLInterfaceType | GraphQLUnionType;

export const builtInScalarMap = {
  [GraphQLString.name]: 'string',
  [GraphQLInt.name]: 'number',
  [GraphQLFloat.name]: 'number',
  [GraphQLBoolean.name]: 'boolean',
  [GraphQLID.name]: 'string',
}

export function printInputGraphQLField(
  name: string,
  type: GraphQLInputType | undefined,
  generator: CodeGenerator,
) {
  if (type === undefined || getRootGraphQLType(type) === undefined) {
    throw new Error(`An invalid type was provided for key '${name}'. This probably means there is a typo in your document or that one of your input types has been renamed.`);
  }

  printPropertyKey(name, !(type instanceof GraphQLNonNull), generator);
  printInputGraphQLType(type, generator);
  generator.print(',');
}

function printInputGraphQLType(
  type: GraphQLInputType,
  generator: CodeGenerator,
) {
  let nullable = true;

  if (type instanceof GraphQLNonNull) {
    nullable = false;
    type = type.ofType;
  }
  
  if (type instanceof GraphQLList) {
    const subType = type.ofType;

    if (subType instanceof GraphQLNonNull) {
      printInputGraphQLType(type.ofType, generator);
    } else {
      generator.print('(');
      printInputGraphQLType(type.ofType, generator);
      generator.print(')');
    }

    generator.print('[]');
  } else if (builtInScalarMap.hasOwnProperty((type as any).name)) {
    generator.print(builtInScalarMap[(type as any).name]);
  } else {
    generator.print(`Schema.${(type as any).name}`);
  }

  generator.print(nullable && ' | null');
}

export function printRootGraphQLType(
  type: GraphQLType,
  generator: CodeGenerator,
) {
  if (type instanceof GraphQLInputObjectType) {
    printInterface({name: type.name}, () => {
      const fields = (type as GraphQLInputObjectType).getFields();
      Object.keys(fields).forEach((name) => {
        const field = fields[name];
        printInputGraphQLField(name, field.type, generator);
      });
    }, generator);
  } else if (type instanceof GraphQLEnumType) {
    const values = type.getValues();
    generator.print(`type ${type.name} = `);
    values.forEach((value, index) => {
      generator.print(`'${value.value}'`);
      generator.print(index !== values.length - 1 && ' | ');
    });
    generator.print(';');
  } else if (type instanceof GraphQLScalarType) {
    generator.print(`type ${type.name} = string;`);
  }
}

export function getRootGraphQLType(type: GraphQLType): GraphQLPrintableType {
  let finalType = type;

  while (finalType instanceof GraphQLNonNull || finalType instanceof GraphQLList) {
    finalType = finalType.ofType;
  }

  return finalType as GraphQLPrintableType;
}
