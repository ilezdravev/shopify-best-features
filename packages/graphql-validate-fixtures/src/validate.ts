import {dirname, basename} from 'path';
import {
  GraphQLType,
  GraphQLScalarType,
  GraphQLEnumType,
  GraphQLNonNull,
  GraphQLObjectType,
  GraphQLList,
  GraphQLString,
  GraphQLInt,
  GraphQLFloat,
  GraphQLBoolean,
} from 'graphql';
import {AST, Field} from 'graphql-tool-utilities/ast';

export type KeyPath = string;

export interface Fixture {
  path: string,
  content: any,
}

export interface Error {
  keyPath?: KeyPath,
  message: string,
}

export interface Validation {
  operationName: string,
  operationType: 'query' | 'mutation',
  operationPath?: string,
  errors: Error[],
}

const OPERATION_MARKER = '@operation';

export default function validateFixtureAgainstAST(fixture: Fixture, ast: AST): Validation {
  const fixtureDirectoryName = basename(dirname(fixture.path));
  const operationName = fixture.content[OPERATION_MARKER] || fixtureDirectoryName;
  const operation = ast.operations[operationName];

  if (operation == null) {
    let lookedFor = `'${fixtureDirectoryName}' based on the fixtureâ€™s directory name`
    
    if (fixture.content[OPERATION_MARKER]) {
      lookedFor += `, and '${fixture.content[OPERATION_MARKER]}' based on the '${OPERATION_MARKER}' key`;
    }

    throw new Error([
      `Could not find a matching operation (looked for ${lookedFor}).`,
      `Make sure to put your fixture in a folder named the same as the operation, or add an '${OPERATION_MARKER}' key indicating the operation.`,
      `Available operations: ${Object.keys(ast.operations).join(', ')}`
    ].join(' '));
  }

  const {fields, filePath, operationType} = operation;
  const value = {...fixture.content};
  delete value[OPERATION_MARKER];

  return {
    operationName,
    operationType,
    operationPath: filePath,
    errors: fields.reduce((allErrors: Error[], field) => {
      return allErrors.concat(
        validateValueAgainstFieldDescription(value[field.responseName], field, '', ast)
      );
    }, []),
  };
}

function validateValueAgainstFieldDescription(value: any, fieldDescription: Field, parentKeyPath: string, ast: AST): Error[] {
  const {type, responseName} = fieldDescription;
  const keyPath = updateKeyPath(parentKeyPath, responseName);
  const typeErrors = validateValueAgainstType(value, type, keyPath);

  if (typeErrors.length > 0) {
    return typeErrors;
  }
  
  return Array.isArray(value)
    ? validateListAgainstFieldDescription(value, fieldDescription, fieldDescription.type, keyPath, ast)
    : validateValueAgainstObjectFieldDescription(value, fieldDescription, keyPath, ast);
}

function validateValueAgainstObjectFieldDescription(value: any, fieldDescription: Field, keyPath: string, ast: AST) {
  if (typeof value !== 'object' || Array.isArray(value)) {
    return [];
  }

  const {fields = [], fragmentSpreads = [], type} = fieldDescription;

  let fragmentFields: Field[] = [];

  if (fragmentSpreads) {
    fragmentSpreads
      .map((spread) => ast.fragments[spread])
      .forEach((fragment) => {
        fragment.fields.forEach((field) => {
          if (fields.some(({fieldName}) => fieldName === field.fieldName)) {
            return;
          }

          const isGuaranteedTypeMatch = fragment.possibleTypes.includes(makeTypeNullable(type));

          fragmentFields.push(
            isGuaranteedTypeMatch
              ? field
              : {...field, type: makeTypeNullable(field.type)}
          );
        });
      });
  }

  return validateValueAgainstFields(value, fields.concat(fragmentFields), keyPath, ast);
}

function makeTypeNullable(type: GraphQLType) {
  return type instanceof GraphQLNonNull
    ? type.ofType
    : type;
}

function validateListAgainstFieldDescription(value: any[], fieldDescription: Field, type: GraphQLType, keyPath: string, ast: AST): Error[] {
  const itemType = type instanceof GraphQLNonNull
      ? (type as GraphQLNonNull<GraphQLList<GraphQLType>>).ofType.ofType
      : (type as GraphQLList<GraphQLType>).ofType;

  return value.reduce((allErrors, item, index) => {
    const itemKeyPath = updateKeyPath(keyPath, index);
    const itemTypeErrors = validateValueAgainstType(item, itemType, itemKeyPath);

    if (itemTypeErrors.length > 0) {
      return allErrors.concat(itemTypeErrors);
    }

    return Array.isArray(item)
      ? allErrors.concat(validateListAgainstFieldDescription(item, fieldDescription, itemType, itemKeyPath, ast))
      : allErrors.concat(validateValueAgainstObjectFieldDescription(item, fieldDescription, itemKeyPath, ast));
  }, []);
}

function validateValueAgainstFields(value: {[key: string]: any}, fields: Field[], keyPath: KeyPath, ast: AST) {
  const finalValue = value || {};

  const excessFields = Object
    .keys(finalValue)
    .filter((key) => fields.every(({responseName}) => key !== responseName))
    .map((key) => error(keyPath, `has key '${key}' not present in the query`));

  return fields.reduce((allErrors, field) => {
    return allErrors.concat(
      validateValueAgainstFieldDescription(finalValue[field.responseName], field, keyPath, ast)
    );
  }, excessFields);
}

function validateValueAgainstType(value: any, type: GraphQLType, keyPath: KeyPath): Error[] {
  if (type instanceof GraphQLNonNull) {
    return value == null
      ? [error(keyPath, `should be non-null but was ${String(value)}`)]
      : validateValueAgainstType(value, type.ofType, keyPath);
  }

  if (value == null) { return []; }

  const valueType = typeof value;

  if (type instanceof GraphQLObjectType) {
    return valueType === 'object'
      ? []
      : [error(keyPath, `should be an object but was a ${valueType}`)];
  }

  if (type instanceof GraphQLList) {
    if (!Array.isArray(value)) {
      return [error(keyPath, `should be an array, but was a ${valueType}`)];
    }

    return Array.isArray(value)
      ? []
      : [error(keyPath, `should be an array, but was a ${valueType}`)];
  }

  if (type === GraphQLString) {
    return valueType === 'string'
      ? []
      : [error(keyPath, `should be a string but was a ${valueType}`)];
  } else if (type === GraphQLInt) {
    if (typeof value === 'number') {
      return Number.isInteger(value)
        ? []
        : [error(keyPath, 'should be an integer but was a float')]
    }

    return [error(keyPath, `should be an integer but was a ${valueType}`)];
  } else if (type === GraphQLFloat) {
    if (typeof value === 'number') {
      return Number.isInteger(value)
        ? [error(keyPath, 'should be a float but was an integer')]
        : [];
    }

    return [error(keyPath, `should be a float but was a ${valueType}`)];
  } else if (type === GraphQLBoolean) {
    return typeof value === 'boolean'
      ? []
      : [error(keyPath, `should be a boolean but was a ${valueType}`)];
  }

  if (type instanceof GraphQLScalarType) {
    return type.parseValue(value) != null
      ? []
      : [error(keyPath, `value does not match scalar ${type.name}`)];
  }

  if (type instanceof GraphQLEnumType) {
    return type.parseValue(value) != null
      ? []
      : [error(keyPath, `value does not match enum ${type.name} (available values: ${type.getValues().map((enumValue) => enumValue.value).join(', ')})`)];
  }

  return [];
}

function updateKeyPath(keyPath: KeyPath, newKey: string | number) {
  if (typeof newKey === 'number') {
    return `${keyPath}[${newKey}]`;
  }

  return keyPath ? `${keyPath}.${newKey}` : newKey;
}

function error(keyPath: KeyPath, message: string): Error {
  return {keyPath, message};
}
