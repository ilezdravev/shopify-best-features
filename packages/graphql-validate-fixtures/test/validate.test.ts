import {buildSchema, parse, concatAST} from 'graphql';
import {compile} from 'graphql-tool-utilities';
import validate from '../validate';

describe('validate()', () => {
  it('validates strings', () => {
    const document = createDocument(`
      type Query { name: String! }
      schema { query: Query }
    `, 'query NameQuery { name }');

    expect(validate({name: 42}, document.operations.NameQuery, document)).toMatchSnapshot();
    expect(validate({}, document.operations.NameQuery, document)).toMatchSnapshot();
    expect(validate({name: null}, document.operations.NameQuery, document)).toMatchSnapshot();
    expect(validate({name: 'Chris'}, document.operations.NameQuery, document)).toMatchSnapshot();

    const document = createDocument(`
      type Query { address: String }
      schema { query: Query }
    `, 'query AddressQuery { address }');

    expect(validate({address: 42}, document.operations.AddressQuery, document)).toMatchSnapshot();
    expect(validate({}, document.operations.AddressQuery, document)).toMatchSnapshot();
    expect(validate({address: null}, document.operations.AddressQuery, document)).toMatchSnapshot();
    expect(validate({address: 'Chris'}, document.operations.AddressQuery, document)).toMatchSnapshot();
  });

  it('validates numbers', () => {
    const document = createDocument(`
      type Query { age: Int! }
      schema { query: Query }
    `, 'query AgeQuery { age }');

    expect(validate({age: '42'}, document.operations.AgeQuery, document)).toMatchSnapshot();
    expect(validate({}, document.operations.AgeQuery, document)).toMatchSnapshot();
    expect(validate({age: null}, document.operations.AgeQuery, document)).toMatchSnapshot();
    expect(validate({age: 42}, document.operations.AgeQuery, document)).toMatchSnapshot();
    expect(validate({age: 42.5}, document.operations.AgeQuery, document)).toMatchSnapshot();

    const document = createDocument(`
      type Query { height: Float }
      schema { query: Query }
    `, 'query HeightQuery { height }');

    expect(validate({height: '42.3'}, document.operations.HeightQuery, document)).toMatchSnapshot();
    expect(validate({}, document.operations.HeightQuery, document)).toMatchSnapshot();
    expect(validate({height: null}, document.operations.HeightQuery, document)).toMatchSnapshot();
    expect(validate({height: 42.3}, document.operations.HeightQuery, document)).toMatchSnapshot();
    expect(validate({height: 42}, document.operations.HeightQuery, document)).toMatchSnapshot();
  });

  it('validates booleans', () => {
    const document = createDocument(`
      type Query { married: Boolean! }
      schema { query: Query }
    `, 'query MarriedQuery { married }');

    expect(validate({married: 'true'}, document.operations.MarriedQuery, document)).toMatchSnapshot();
    expect(validate({}, document.operations.MarriedQuery, document)).toMatchSnapshot();
    expect(validate({married: null}, document.operations.MarriedQuery, document)).toMatchSnapshot();
    expect(validate({married: true}, document.operations.MarriedQuery, document)).toMatchSnapshot();
    expect(validate({married: false}, document.operations.MarriedQuery, document)).toMatchSnapshot();

    const document = createDocument(`
      type Query { responded: Boolean }
      schema { query: Query }
    `, 'query RespondedQuery { responded }');

    expect(validate({responded: 'true'}, document.operations.RespondedQuery, document)).toMatchSnapshot();
    expect(validate({}, document.operations.RespondedQuery, document)).toMatchSnapshot();
    expect(validate({responded: null}, document.operations.RespondedQuery, document)).toMatchSnapshot();
    expect(validate({responded: true}, document.operations.RespondedQuery, document)).toMatchSnapshot();
    expect(validate({responded: false}, document.operations.RespondedQuery, document)).toMatchSnapshot();
  });

  describe('lists', () => {
    it('validates the types of lists', () => {
      const document = createDocument(`
        type Query { favoriteNumbers: [Int!]! }
        schema { query: Query }
      `, 'query FavoriteNumbersQuery { favoriteNumbers }');

      expect(validate({}, document.operations.FavoriteNumbersQuery, document)).toMatchSnapshot();
      expect(validate({favoriteNumbers: null}, document.operations.FavoriteNumbersQuery, document)).toMatchSnapshot();
      expect(validate({favoriteNumbers: []}, document.operations.FavoriteNumbersQuery, document)).toMatchSnapshot();
      expect(validate({favoriteNumbers: {}}, document.operations.FavoriteNumbersQuery, document)).toMatchSnapshot();
      expect(validate({favoriteNumbers: [1, 2.5, null, true, 5]}, document.operations.FavoriteNumbersQuery, document)).toMatchSnapshot();
    });

    it('validates nested arrays', () => {
      const document = createDocument(`
        type Query { sets: [[Int!]!] }
        schema { query: Query }
      `, 'query SetsQuery { sets }');

      expect(validate({}, document.operations.SetsQuery, document)).toMatchSnapshot();
      expect(validate({sets: null}, document.operations.SetsQuery, document)).toMatchSnapshot();
      expect(validate({sets: []}, document.operations.SetsQuery, document)).toMatchSnapshot();
      expect(validate({sets: [1, 2, 3]}, document.operations.SetsQuery, document)).toMatchSnapshot();
      expect(validate({sets: [[1, 2, null], [true, {}, 3]]}, document.operations.SetsQuery, document)).toMatchSnapshot();
      expect(validate({sets: [null, [1, 2], true]}, document.operations.SetsQuery, document)).toMatchSnapshot();
    });

    it('validates objects nested in arrays', () => {
      const document = createDocument(`
        type Person {
          name: String!
          age: Int!
        }
        type Query { people: [Person!]! }
        schema { query: Query }
      `, 'query PeopleQuery { people { name } }');

      expect(validate({}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: null}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: []}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: [1]}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: [{}]}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: [{name: true}]}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: [{name: 'Chris'}]}, document.operations.PeopleQuery, document)).toMatchSnapshot();
      expect(validate({people: [{name: 'Chris', age: 42}]}, document.operations.PeopleQuery, document)).toMatchSnapshot();
    });
  });

  describe('objects', () => {
    it('validates that the value is actually an object', () => {
      const document = createDocument(`
        type Person { name: String! }
        type Query { person: Person! }
        schema { query: Query }
      `, 'query PersonQuery { person { name } }');

      expect(validate({}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: null}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: true}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: 'Chris'}}, document.operations.PersonQuery, document)).toMatchSnapshot();
    });

    it('creates errors for keys not present in the query', () => {
      const document = createDocument(`
        type Person {
          name: String!
          age: Int
        }
        type Query { person: Person! }
        schema { query: Query }
      `, 'query PersonQuery { person { name } }');

      expect(validate({person: {name: 'Chris', age: 42}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: 'Chris', foo: true}}, document.operations.PersonQuery, document)).toMatchSnapshot();
    });

    it('validates nested objects', () => {
      const document = createDocument(`
        type Address { number: Int }
        type Person { address: Address! }
        type Query { person: Person! }
        schema { query: Query }
      `, 'query AddressQuery { person { address { number } } }');

      expect(validate({}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: null}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: {}}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: {address: {}}}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: {address: {number: '63'}}}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: {address: {number: 15.4}}}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: {address: {number: 63}}}, document.operations.AddressQuery, document)).toMatchSnapshot();
      expect(validate({person: {address: {street: 'Main st'}}}, document.operations.AddressQuery, document)).toMatchSnapshot();
    });
  });

  describe('fragments', () => {
    it('includes fields from inline fragments', () => {
      const document = createDocument(`
        type Person { name: String! }
        type Query { person: Person! }
        schema { query: Query }
      `, 'query PersonQuery { person { ... on Person { name } } }');

      expect(validate({}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: null}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: true}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: 'Chris'}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: 'Chris', age: 42}}, document.operations.PersonQuery, document)).toMatchSnapshot();
    });

    it('includes fields from separate fragments', () => {
      const document = createDocument(`
        type Person { name: String! }
        type Query { person: Person! }
        schema { query: Query }
      `, 'query PersonQuery { person { ...PersonFragment } }', 'fragment PersonFragment on Person { name }');

      expect(validate({}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: null}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: true}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: 'Chris'}}, document.operations.PersonQuery, document)).toMatchSnapshot();
      expect(validate({person: {name: 'Chris', age: 42}}, document.operations.PersonQuery, document)).toMatchSnapshot();
    });

    it('treats fields on interface type spreads to be nullable', () => {
      const document = createDocument(`
        interface Animal { name: String! }
        type Cat implements Animal {
          name: String!
          purrs: Boolean!
        }
        type Dog implements Animal {
          name: String!
          barks: Boolean!
        }
        type Query { animals: [Animal!]! }
        schema { query: Query }
      `, 'query DogQuery { animals { name, ...DogFragment } }', 'fragment DogFragment on Dog { barks }');

      expect(validate({}, document.operations.DogQuery, document)).toMatchSnapshot();
      expect(validate({animals: null}, document.operations.DogQuery, document)).toMatchSnapshot();
      expect(validate({animals: []}, document.operations.DogQuery, document)).toMatchSnapshot();
      expect(validate({animals: [{}]}, document.operations.DogQuery, document)).toMatchSnapshot();
      expect(validate({animals: [{name: 'Winston'}]}, document.operations.DogQuery, document)).toMatchSnapshot();
      expect(validate({animals: [{name: 'Winston', barks: 40}]}, document.operations.DogQuery, document)).toMatchSnapshot();
      expect(validate({animals: [{name: 'Winston', barks: true}]}, document.operations.DogQuery, document)).toMatchSnapshot();
    });
  });

  // TODO: multiple top-level items
  // TODO: aliases
});

function createDocument(schemaString: string, ...queryStrings: string[]) {
  const schema = buildSchema(schemaString);
  const ast = concatAST(queryStrings.map((queryString) => parse(queryString)));
  return compile(schema, ast);
}
