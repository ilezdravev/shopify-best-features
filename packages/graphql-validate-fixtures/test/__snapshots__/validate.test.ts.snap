// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validate() fragments includes fields from inline fragments 1`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() fragments includes fields from inline fragments 2`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() fragments includes fields from inline fragments 3`] = `
Array [
  Object {
    "keyPath": "person.name",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() fragments includes fields from inline fragments 4`] = `
Array [
  Object {
    "keyPath": "person.name",
    "message": "should be a string but was a boolean",
  },
]
`;

exports[`validate() fragments includes fields from inline fragments 5`] = `Array []`;

exports[`validate() fragments includes fields from inline fragments 6`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "has key 'age' not present in the query",
  },
]
`;

exports[`validate() fragments includes fields from separate fragments 1`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() fragments includes fields from separate fragments 2`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() fragments includes fields from separate fragments 3`] = `
Array [
  Object {
    "keyPath": "person.name",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() fragments includes fields from separate fragments 4`] = `
Array [
  Object {
    "keyPath": "person.name",
    "message": "should be a string but was a boolean",
  },
]
`;

exports[`validate() fragments includes fields from separate fragments 5`] = `Array []`;

exports[`validate() fragments includes fields from separate fragments 6`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "has key 'age' not present in the query",
  },
]
`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 1`] = `
Array [
  Object {
    "keyPath": "animals",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 2`] = `
Array [
  Object {
    "keyPath": "animals",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 3`] = `Array []`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 4`] = `
Array [
  Object {
    "keyPath": "animals[0].name",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 5`] = `Array []`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 6`] = `
Array [
  Object {
    "keyPath": "animals[0].barks",
    "message": "should be a boolean but was a number",
  },
]
`;

exports[`validate() fragments treats fields on interface type spreads to be nullable 7`] = `Array []`;

exports[`validate() lists validates nested arrays 1`] = `Array []`;

exports[`validate() lists validates nested arrays 2`] = `Array []`;

exports[`validate() lists validates nested arrays 3`] = `Array []`;

exports[`validate() lists validates nested arrays 4`] = `
Array [
  Object {
    "keyPath": "sets[0]",
    "message": "should be an array, but was a number",
  },
  Object {
    "keyPath": "sets[1]",
    "message": "should be an array, but was a number",
  },
  Object {
    "keyPath": "sets[2]",
    "message": "should be an array, but was a number",
  },
]
`;

exports[`validate() lists validates nested arrays 5`] = `
Array [
  Object {
    "keyPath": "sets[0][2]",
    "message": "should be non-null but was null",
  },
  Object {
    "keyPath": "sets[1][0]",
    "message": "should be an integer but was a boolean",
  },
  Object {
    "keyPath": "sets[1][1]",
    "message": "should be an integer but was a object",
  },
]
`;

exports[`validate() lists validates nested arrays 6`] = `
Array [
  Object {
    "keyPath": "sets[0]",
    "message": "should be non-null but was null",
  },
  Object {
    "keyPath": "sets[2]",
    "message": "should be an array, but was a boolean",
  },
]
`;

exports[`validate() lists validates objects nested in arrays 1`] = `
Array [
  Object {
    "keyPath": "people",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() lists validates objects nested in arrays 2`] = `
Array [
  Object {
    "keyPath": "people",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() lists validates objects nested in arrays 3`] = `Array []`;

exports[`validate() lists validates objects nested in arrays 4`] = `
Array [
  Object {
    "keyPath": "people[0]",
    "message": "should be an object but was a number",
  },
]
`;

exports[`validate() lists validates objects nested in arrays 5`] = `
Array [
  Object {
    "keyPath": "people[0].name",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() lists validates objects nested in arrays 6`] = `
Array [
  Object {
    "keyPath": "people[0].name",
    "message": "should be a string but was a boolean",
  },
]
`;

exports[`validate() lists validates objects nested in arrays 7`] = `Array []`;

exports[`validate() lists validates objects nested in arrays 8`] = `
Array [
  Object {
    "keyPath": "people[0]",
    "message": "has key 'age' not present in the query",
  },
]
`;

exports[`validate() lists validates the types of lists 1`] = `
Array [
  Object {
    "keyPath": "favoriteNumbers",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() lists validates the types of lists 2`] = `
Array [
  Object {
    "keyPath": "favoriteNumbers",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() lists validates the types of lists 3`] = `Array []`;

exports[`validate() lists validates the types of lists 4`] = `
Array [
  Object {
    "keyPath": "favoriteNumbers",
    "message": "should be an array, but was a object",
  },
]
`;

exports[`validate() lists validates the types of lists 5`] = `
Array [
  Object {
    "keyPath": "favoriteNumbers[1]",
    "message": "should be an integer but was a float",
  },
  Object {
    "keyPath": "favoriteNumbers[2]",
    "message": "should be non-null but was null",
  },
  Object {
    "keyPath": "favoriteNumbers[3]",
    "message": "should be an integer but was a boolean",
  },
]
`;

exports[`validate() objects creates errors for keys not present in the query 1`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "has key 'age' not present in the query",
  },
]
`;

exports[`validate() objects creates errors for keys not present in the query 2`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "has key 'foo' not present in the query",
  },
]
`;

exports[`validate() objects validates nested objects 1`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() objects validates nested objects 2`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() objects validates nested objects 3`] = `
Array [
  Object {
    "keyPath": "person.address",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() objects validates nested objects 4`] = `Array []`;

exports[`validate() objects validates nested objects 5`] = `
Array [
  Object {
    "keyPath": "person.address.number",
    "message": "should be an integer but was a string",
  },
]
`;

exports[`validate() objects validates nested objects 6`] = `
Array [
  Object {
    "keyPath": "person.address.number",
    "message": "should be an integer but was a float",
  },
]
`;

exports[`validate() objects validates nested objects 7`] = `Array []`;

exports[`validate() objects validates nested objects 8`] = `
Array [
  Object {
    "keyPath": "person.address",
    "message": "has key 'street' not present in the query",
  },
]
`;

exports[`validate() objects validates that the value is actually an object 1`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() objects validates that the value is actually an object 2`] = `
Array [
  Object {
    "keyPath": "person",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() objects validates that the value is actually an object 3`] = `
Array [
  Object {
    "keyPath": "person.name",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() objects validates that the value is actually an object 4`] = `
Array [
  Object {
    "keyPath": "person.name",
    "message": "should be a string but was a boolean",
  },
]
`;

exports[`validate() objects validates that the value is actually an object 5`] = `Array []`;

exports[`validate() validates booleans 1`] = `
Array [
  Object {
    "keyPath": "married",
    "message": "should be a boolean but was a string",
  },
]
`;

exports[`validate() validates booleans 2`] = `
Array [
  Object {
    "keyPath": "married",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() validates booleans 3`] = `
Array [
  Object {
    "keyPath": "married",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() validates booleans 4`] = `Array []`;

exports[`validate() validates booleans 5`] = `Array []`;

exports[`validate() validates booleans 6`] = `
Array [
  Object {
    "keyPath": "responded",
    "message": "should be a boolean but was a string",
  },
]
`;

exports[`validate() validates booleans 7`] = `Array []`;

exports[`validate() validates booleans 8`] = `Array []`;

exports[`validate() validates booleans 9`] = `Array []`;

exports[`validate() validates booleans 10`] = `Array []`;

exports[`validate() validates numbers 1`] = `
Array [
  Object {
    "keyPath": "age",
    "message": "should be an integer but was a string",
  },
]
`;

exports[`validate() validates numbers 2`] = `
Array [
  Object {
    "keyPath": "age",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() validates numbers 3`] = `
Array [
  Object {
    "keyPath": "age",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() validates numbers 4`] = `Array []`;

exports[`validate() validates numbers 5`] = `
Array [
  Object {
    "keyPath": "age",
    "message": "should be an integer but was a float",
  },
]
`;

exports[`validate() validates numbers 6`] = `
Array [
  Object {
    "keyPath": "height",
    "message": "should be a float but was a string",
  },
]
`;

exports[`validate() validates numbers 7`] = `Array []`;

exports[`validate() validates numbers 8`] = `Array []`;

exports[`validate() validates numbers 9`] = `Array []`;

exports[`validate() validates numbers 10`] = `
Array [
  Object {
    "keyPath": "height",
    "message": "should be a float but was an integer",
  },
]
`;

exports[`validate() validates strings 1`] = `
Array [
  Object {
    "keyPath": "name",
    "message": "should be a string but was a number",
  },
]
`;

exports[`validate() validates strings 2`] = `
Array [
  Object {
    "keyPath": "name",
    "message": "should be non-null but was undefined",
  },
]
`;

exports[`validate() validates strings 3`] = `
Array [
  Object {
    "keyPath": "name",
    "message": "should be non-null but was null",
  },
]
`;

exports[`validate() validates strings 4`] = `Array []`;

exports[`validate() validates strings 5`] = `
Array [
  Object {
    "keyPath": "address",
    "message": "should be a string but was a number",
  },
]
`;

exports[`validate() validates strings 6`] = `Array []`;

exports[`validate() validates strings 7`] = `Array []`;

exports[`validate() validates strings 8`] = `Array []`;
